integration_tests/src/main/python/data_gen.py

```diff
@@ -613,31 +613,112 @@ def make_null():
         self._start(rand, make_null)
 
 # DayTimeIntervalGen is for Spark 3.3.0+
-# DayTimeIntervalType(startField, endField): Represents a day-time interval which is made up of a contiguous subset of the following fields:
+# DayTimeIntervalType(startField, endField):
+# Represents a day-time interval which is made up of a contiguous subset of the following fields:
 #   SECOND, seconds within minutes and possibly fractions of a second [0..59.999999],
+#   Note Spark now uses 99 as max second, see issue https://issues.apache.org/jira/browse/SPARK-38324
+#   If second is start field, it's max value is long.max / microseconds in one second
 #   MINUTE, minutes within hours [0..59],
+#   If minute is start field, it's max value is long.max / microseconds in one minute
 #   HOUR, hours within days [0..23],
-#   DAY, days in the range [0..106751991].
+#   If hour is start field, it's max value is long.max / microseconds in one hour
+#   DAY, days in the range [0..106751991]. 106751991 is long.max / microseconds in one day
 # For more details: https://spark.apache.org/docs/latest/sql-ref-datatypes.html
-# Note: 106751991/365 = 292471 years which is much bigger than 9999 year, seems something is wrong
+# Note: 106751991/365 = 292471 years which is much bigger than 9999 year
 class DayTimeIntervalGen(DataGen):
     """Generate DayTimeIntervalType values"""
-    def __init__(self, max_days = None, nullable=True, special_cases =[timedelta(seconds = 0)]):
+    def __init__(self, max_days=None, start_field="day", end_field="second", allow_negative=True, nullable=True,
+                 special_cases=[timedelta(seconds=0)]):
         super().__init__(DayTimeIntervalType(), nullable=nullable, special_cases=special_cases)
         if max_days is None:
             self._max_days = 106751991
         else:
+            assert 106751991 >= max_days > 0
             self._max_days = max_days
+        self._allow_negative = allow_negative
+        self._start_field = start_field
+        self._end_field = end_field
+
+        fields = ["day", "hour", "minute", "second"]
+        start_index = fields.index(start_field)
+        end_index = fields.index(end_field)
+        if start_index > end_index:
+            raise RuntimeError('Start field {}, end field {}, valid fields is {}, start field should <= end field'.format(start_field, end_field, fields))
+
+        super().__init__(DayTimeIntervalType(start_index, end_index), nullable=nullable, special_cases=special_cases)
+
+    def _gen_random(self, rand, start_field, end_field):
+        micros_per_second = 1000 * 1000
+        micros_per_minute = 60 * micros_per_second
+        micros_per_hour = 60 * micros_per_minute
+        micros_per_day = 24 * micros_per_hour
+
+        max_micros = self._max_days * micros_per_day
+
+        # set default value
+        days = 0
+        hours = 0
+        minutes = 0
+        seconds = 0
+        microseconds = 0
+
+        if (start_field, end_field) == ("day", "day"):
```

Comment:
I think it would be better to use the below structure, which should get a better perf.
```
if:
    ...
elif:
    ...
elif:
    ...
else:
    ...
```

Reply:
Can we do this totally differently? There is just so much copy/paste between each part of the code. We know that the python code really is only converting it to micro seconds. So can we just generate a random number for the microseconds, up to `max_micros`, and then truncate it to the proper time? Also can we convert the "string" start/end fields into something simpler to write code for?

```
# in __init__
DAY = 0
HOUR = 1
MIN = 2
SEC = 3
fields_to_look_at = ["day", "hour", "minute", "second"]
si = fields_to_look_at.index(start_field)
ei = fileds_to_look_at.index(end_field)
assert si <= ei

self.hasDays = si <= DAY and ei >= DAY
self.hasHours = si <= HOUR and ei >= HOUR
self.hasMin = si <= MIN and ei >= MIN
self.hasSec = si <= SEC and ei >= SEC
```